src/storage/src/hummock/store/version.rs

```diff
@@ -456,32 +415,50 @@ impl HummockReadVersion {
     }
 
     pub fn add_merged_imm(&mut self, merged_imm: ImmutableMemtable) {
-        let staging_imm_count = self.staging.imm.len();
-        let merged_imm_ids = merged_imm.get_imm_ids();
-
-        #[cfg(debug_assertions)]
-        {
-            // check the suffix `merged_imm_ids.len()` imms in staging.imm are the same as
-            // `merged_imm_ids`
-            let diff = staging_imm_count - merged_imm_ids.len();
-            let mut count: usize = 0;
-            for (i, imm) in self.staging.imm.iter().skip(diff).enumerate() {
-                count += 1;
-                assert_eq!(
-                    imm.batch_id(),
-                    merged_imm_ids[i],
-                    "merged_imm_ids: {:?}",
-                    merged_imm_ids
-                );
+        assert!(merged_imm.get_imm_ids().iter().rev().is_sorted());
+        let min_imm_id = *merged_imm.get_imm_ids().last().expect("non-empty");
+
+        let back = self.staging.imm.back().expect("should not be empty");
+
+        // pop and save imms that are written earlier than the oldest imm if there is any
+        let earlier_imms = if back.batch_id() < min_imm_id {
```

Comment:
Why not just `retain` imms that are not in the merged_imm? Which will not break the original ordering of existing imms. e.g.
```rust
self.staging
            .imm
            .retain(|imm| !merged_imm_ids.contains(&imm.batch_id()))
```

Reply:
Yes, retain works, but we should have some sanity check to ensure that the input and output is expected, and otherwise the query result will be incorrect.